Lib/idlelib/idle_test/test_editor.py

```diff
@@ -1,14 +1,432 @@
+""" Test idlelib.editor.
+"""
+
 import unittest
-from idlelib.editor import EditorWindow
+import tkinter as tk
+import sys
+from functools import partial
+from idlelib import editor
+from idlelib.multicall import MultiCallCreator
+from test.support import requires
+from unittest import mock
+
+root = None
+editwin = None
+
+
+def setUpModule():
+    global root, editwin
+    requires('gui')
+    root = tk.Tk()
+    root.withdraw()
+    editwin = editor.EditorWindow(root=root)
+
+
+def tearDownModule():
+    global root, editwin
+    editwin.close()
+    del editwin
+    root.update_idletasks()
+    root.destroy()
+    del root
+
 
 class Editor_func_test(unittest.TestCase):
     def test_filename_to_unicode(self):
-        func = EditorWindow._filename_to_unicode
+        func = editor.EditorWindow._filename_to_unicode
         class dummy(): filesystemencoding = 'utf-8'
         pairs = (('abc', 'abc'), ('a\U00011111c', 'a\ufffdc'),
                  (b'abc', 'abc'), (b'a\xf0\x91\x84\x91c', 'a\ufffdc'))
         for inp, out in pairs:
             self.assertEqual(func(dummy, inp), out)
 
+
+class ModuleHelpersTest(unittest.TestCase):
+    """Test functions defined at the module level."""
+
+    def test_prepstr(self):
+        ps = editor.prepstr
+        eq = self.assertEqual
+        eq(ps('_spam'), (0, 'spam'))
+        eq(ps('spam'), (-1, 'spam'))
+        eq(ps('spam_'), (4, 'spam'))
+
+    @mock.patch.object(editor.macosx, 'isCocoaTk')
+    def test_get_accelerator(self, mock_cocoa):
+        ga = editor.get_accelerator
+        eq = self.assertEqual
+        keydefs = {'<<zoom-height>>': ['<Alt-Key-9>'],
+                   '<<open-module>>': ['<Control-Shift-O>'],
+                   '<<cancel>>': ['<Cancel>'],
+                   '<<indent>>': ['<Alt-bracketleft>'],
+                   '<<copy>>': ['<Control-j>', '<Control-c>']}
+
+        mock_cocoa.return_value = False
+        eq(ga(keydefs, '<<paste>>'), '')  # Not in keydefs.
+        eq(ga(keydefs, '<<copy>>'), 'Ctrl+J')  # Control to Ctrl and first only.
+        eq(ga(keydefs, '<<zoom-height>>'), 'Alt+9')  # Remove Key-.
+        eq(ga(keydefs, '<<indent>>'), 'Alt+[')  # bracketleft to [.
+        eq(ga(keydefs, '<<cancel>>'), 'Ctrl+Break')  # Cancel to Ctrl-Break.
+        eq(ga(keydefs, '<<open-module>>'), 'Ctrl+Shift+O')  # Shift doesn't change.
+
+        # Cocoa test.
+        mock_cocoa.return_value = True
+        eq(ga(keydefs, '<<open-module>>'), '')  # Cocoa skips open-module shortcut.
+
+
+class MenubarTest(unittest.TestCase):
+    """Test functions involved with creating the menubar."""
+
+    @classmethod
+    def setUpClass(cls):
+        # Test the functions called during the __init__ for
+        # EditorWindow that create the menubar and submenus.
+        # The class is mocked in order to prevent the functions
+        # from being called automatically.
+        w = cls.mock_editwin = mock.Mock(editor.EditorWindow)
```

Comment:
Do I understand correctly that this uses a mock mostly to avoid all of the setup in EditorWindow.__init__()? That could also be accomplished by patching that method (e.g. with `mock.patch`) or by creating a sub-class.